scheduler/remote_server.py

```diff
@@ -0,0 +1,487 @@
+import random
+import subprocess as sp
+import pickle as pkl
+import matplotlib.pyplot as plt
+import matplotlib.dates as mdates
+
+import remote_server_options as rso
+
+def format_to_atq(dt, experiment, first_event_flag=False):
+    """Turns an experiment line from the scd into a formatted atq command.
+
+    Args:
+        dt (datetime): Datetime of the experiment
+        experiment (str): The experiment to run
+        first_event_flag (bool, optional): Flag to signal whether the experiment is the first to
+        run
+
+    Returns:
+        str: Formatted atq str.
+    """
+    if first_event_flag:
+        cmd_str = "echo '{borealis_path}/start_radar.sh {experiment}' | at now"
+        cmd_str = cmd_str.format(borealis_path=os.environ['BOREALISPATH'],experiment=experiment)
+        cmd_str = dt.strftime(cmd_str)
+    else:
+        cmd_str = "echo '{borealis_path}/start_radar.sh {experiment}' | at -t %Y%m%d%H%M"
+        cmd_str = cmd_str.format(borealis_path=os.environ['BOREALISPATH'],experiment=experiment)
+        cmd_str = dt.strftime(cmd_str)
+
+    return cmd_str
+
+def get_next_month(date):
+        """Finds the datetime of the next month.
+
+        Returns:
+            datetime: datetime object of the next month.
+        """
+
+        counter = 1
+        new_date = date + datetime.timedelta(days=counter)
+        while new_date.month == date.month:
+            counter += 1
+            new_date = date + datetime.timedelta(days=counter)
+
+        return new_date
+
+def plot_timeline(timeline_dict, scd_dir, now):
+    """Plots the timeline to better visualize runtime.
+
+    Args:
+        timeline_dict (dict): A dict of grouped entries.
+        scd_dir (str): The scd directory path.
+        now (datetime): The datetime holding the time of scheduling.
+
+    Returns:
+        (str, str): Paths to the saved plots.
+    """
+    fig, ax = plt.subplots()
+
+    event_labels = []
+    first_date, last_date = None, None
+
+
+    def get_cmap(n, name='hsv'):
+        '''Returns a function that maps each index in 0, 1, ..., n-1 to a distinct
+        RGB color; the keyword argument name must be a standard mpl colormap name.'''
+        return plt.cm.get_cmap(name, n)
+
+    #make random colors
+    cmap = get_cmap(len(timeline_dict.items()))
+    colors = [cmap(i) for i in range(len(timeline_dict.items()))]
+    random.shuffle(colors)
+
+    for i, (_, events) in enumerate(timeline_dict.items()):
+        event_times = []
+        event_label = ""
+
+        color = colors[i]
+        colors_for_line = [color] * len(events)
+        for event in events:
+            time_start = mdates.date2num(event['time'])
+
+            if event['duration'] == '-':
+                td = get_next_month(event['time']) - event['time']
+            else:
+                td = datetime.timedelta(minutes=int(event['duration']))
+
+            time_end = td.total_seconds()/(24 * 60 * 60)
+            event_times.append((time_start, time_end))
```

Comment:
td doesn't look right to me if duration is unspecified. Will it always plot all non-specified durations as going until end of the month? Should be overwritten if another scheduling line appears before the end of the month right?

Reply:
Yes. At this point the only infinite duration line left should be the last line in the schedule.